Improve tidy's license validation logic

UK992 opened the pull request:
Rebased and fixed https://github.com/servo/servo/pull/10721, which is inactive for months.
Fixes https://github.com/servo/servo/issues/10716

r? @larsbergstrom or @edunham

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/12781)

<!-- Reviewable:end -->

Commit:
Improve tidy's license validation logic

fixes https://github.com/servo/servo/issues/10716

I took the lazy way out and hardcoded the size of block we examine for
licenses.

fixes https://github.com/servo/servo/issues/10719

Includes tests for new functionality.

highfive commented:
Heads up! This PR modifies the following files:
 * @wafflespeanut: python/tidy/servo_tidy_tests/test_tidy.py, python/tidy/HISTORY.rst, python/tidy/servo_tidy/tidy.py, python/tidy/servo_tidy/licenseck.py, python/tidy/setup.py, python/tidy/servo_tidy_tests/apache2_license.rs

jdm commented on python/tidy/servo_tidy_tests/test_tidy.py:
```diff
@@ -150,3 +154,7 @@ def test_file_list(self):
 def do_tests():
     suite = unittest.TestLoader().loadTestsFromTestCase(CheckTidiness)
     return 0 if unittest.TextTestRunner(verbosity=2).run(suite).wasSuccessful() else 1
+    unittest.TextTestRunner(verbosity=2).run(suite)
```
This is unreachable code.

UK992 replied:
done!

UK992 committed:
Check for blank line after shebang and grab actual comment block instead of hard-coded one

UK992 commented:
with last commit i added grabbing actual comment block instead of hard coded one, and also added checking for blank line after shebang, so that doesn't messed up license check.

jdm commented:
The travis failure is unrelated to these changes.

aneeshusa commented on python/tidy/servo_tidy/licenseck.py:
```diff
@@ -7,97 +7,14 @@
 # option. This file may not be copied, modified, or distributed
 # except according to those terms.
 
+mpl = "This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0. If a copy of the MPL was not distributed with this file, You can obtain one at http://mozilla.org/MPL/2.0/"
```
Why put these onto a single line? AFAIK the indentation never changes, just the comment character at the start of each line. IMO this form is nicer:
```
mpl = """\
This Source Code...
""",
```

Wafflespeanut replied:
Style Nit: Also, the name could be in uppercase, since it's a constant (even though it doesn't make any difference in python).

UK992 replied:
@aneeshusa i changed it to this type of form, and before matching with header remove all newlines.

aneeshusa commented on python/tidy/servo_tidy/tidy.py:
```diff
@@ -147,13 +145,49 @@ def filter_files(start_dir, only_changed_files, progress):
+
+
+def licensed_mpl(header):
+    if licenseck.mpl in header:
```
We can do `return licenseck.mpl in header` directly.

Wafflespeanut commented on python/tidy/servo_tidy/licenseck.py:
```diff
@@ -7,97 +7,14 @@
 
-# These licenses are valid for use in Servo
-licenses = [
+apache = "Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or http://www.apache.org/licenses/LICENSE-2.0> or the MIT license <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your option. This file may not be copied, modified, or distributed except according to those terms."
```
ditto on the name

Wafflespeanut commented on python/tidy/servo_tidy/licenseck.py:
```diff
@@ -7,97 +7,14 @@
+copyright = [
+    "See the COPYRIGHT file at the top-level directory of this distribution",
+    "See http://rust-lang.org/COPYRIGHT",
+]
```
`copyright` is a built-in function. ditto on uppercase?

Wafflespeanut commented on python/tidy/servo_tidy/tidy.py:
```diff
@@ -147,13 +145,49 @@ def filter_files(start_dir, only_changed_files, progress):
+        yield (1, "missing blank line after shebang")
+
+    blank_lines = 0
+    MAX_BLANK_LINES = 2 if lines[0].startswith("#!") else 1
```
Style nit: This shouldn't be uppercase, since it's a variable.

Wafflespeanut commented on python/tidy/servo_tidy_tests/test_tidy.py:
```diff
@@ -150,3 +159,6 @@ def test_file_list(self):
     return 0 if unittest.TextTestRunner(verbosity=2).run(suite).wasSuccessful() else 1
+
+if __name__ == "__main__":
+    do_tests()
```
Why was this change introduced?

UK992 replied:
don't know, i removed it.

Wafflespeanut commented on python/tidy/HISTORY.rst:
```diff
@@ -1,6 +1,17 @@
 Release History
 ---------------
 
+0.1.0 (2016-04-19)
```
psst... date?

Wafflespeanut commented on python/tidy/servo_tidy/tidy.py:
```diff
@@ -147,13 +145,49 @@ def filter_files(start_dir, only_changed_files, progress):
+
+def licensed_apache(header):
+    if licenseck.apache in header:
+        for copyright in licenseck.copyright:
```
ditto on `copyright`

Wafflespeanut commented on python/tidy/servo_tidy/tidy.py:
```diff
@@ -305,7 +339,7 @@ def check_toml(file_name, lines):
         if line.find("*") != -1:
             yield (idx + 1, "found asterisk instead of minimum version number")
-        for license in licenses_toml:
+        for license in licenseck.licenses_toml:
```
`license` is also a built-in function

aneeshusa commented on python/tidy/servo_tidy/licenseck.py:
```diff
@@ -139,3 +66,4 @@
     'name = "webrender"',
     'name = "webrender_traits"',
 ]
+# noqa: Indicate to flake8 that we do not want to check indentation here
```
We shouldn't need this anymore since we don't have unindented strings inside a list (the old `licenses` variable).

UK992 replied:
removed.

Wafflespeanut commented on python/tidy/servo_tidy/tidy.py:
```diff
@@ -17,12 +17,9 @@
 import subprocess
 import sys
-from licenseck import licenses, licenses_toml, licenses_dep_toml
+import licenseck as licenses
```
Nit: This should be `from licenseck import MPL, ...` (we don't wanna prefer `*` even though we import all the names)

Wafflespeanut commented on python/tidy/servo_tidy/tidy.py:
```diff
@@ -17,12 +17,9 @@
-EMACS_HEADER = "/* -*- Mode:"
-VIM_HEADER = "/* vim:"
-MAX_LICENSE_LINESPAN = max(len(license.splitlines()) for license in licenses)
+COMMENTS = ["// ", "# ", " *", "/* ", "*/ "]
```
We won't be needing `*/` here, would we?

Wafflespeanut commented on python/tidy/servo_tidy/tidy.py:
```diff
@@ -147,13 +143,47 @@ def filter_files(start_dir, only_changed_files, progress):
+
+
+def licensed_mpl(header):
+    return licenses.MPL.replace('\n', ' ') in header
```
Instead of doing a string replacement for every function call, we could just have a global variable like `mpl = MPL.replace('\n', ' ')` or something at the top.

UK992 replied:
I added backslash to every line in licenseck.

Wafflespeanut commented on python/tidy/servo_tidy/tidy.py:
```diff
@@ -147,13 +143,47 @@ def filter_files(start_dir, only_changed_files, progress):
+
+
+def licensed_apache(header):
+    if licenses.APACHE.replace('\n', ' ') in header:
```
ditto on "globalizing"

Wafflespeanut commented on python/tidy/servo_tidy/tidy.py:
```diff
@@ -147,13 +143,47 @@ def filter_files(start_dir, only_changed_files, progress):
+    blank_lines = 0
+    max_blank_lines = 2 if lines[0].startswith("#!") else 1
+    license_block = []
+    for l in lines:
```
Nit: Please have a newline before the loop

Wafflespeanut commented on python/tidy/servo_tidy/tidy.py:
```diff
@@ -147,13 +143,47 @@ def filter_files(start_dir, only_changed_files, progress):
+            break
+        line = uncomment(l)
+        if line is not None:
+            license_block += [line]
```
Nit: `license_block.append(line)`

Wafflespeanut commented:
Thanks for doing this! Only a few nits and minor fixes, and we'll be good to go :)

Wafflespeanut commented on python/tidy/servo_tidy/tidy.py:
```diff
@@ -147,13 +143,47 @@ def filter_files(start_dir, only_changed_files, progress):
+        for c in licenses.COPYRIGHT:
+            if c in header:
+                return True
+    return False
```
could be simply,

``` python
return [c in COPYRIGHT if c in header]
```

Ms2ger replied:
I disagree about silently returning `None`; it makes more sense to always return a boolean.

As for the array comprehension... I suppose you meant `return any(c in header for c in COPYRIGHT)`.

Wafflespeanut replied:
That works too. I'm cool with that :)

Wafflespeanut commented:
Looking great! Can you please squash your commits appropriately? Then, we'll land this :)

UK992 commented:
done!

Wafflespeanut commented:
Thanks! :)

@bors-servo r+

bors-servo commented:
:pushpin: Commit 9a3b61a has been approved by `Wafflespeanut`

<!-- @bors-servo r=Wafflespeanut 9a3b61abe1ad633ab723ca6cbb7180a4992ead5a -->

bors-servo commented:
:hourglass: Testing commit 9a3b61a with merge e646469...

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/2579)

highfive commented:
▶ FAIL [expected PASS] /css-transforms-1_dev/html/transform-abspos-007.htm
      └   → /css-transforms-1_dev/html/transform-abspos-007.htm 71f0313eedfbfcce0b0fcc5ae55f34b1daa3b8d8
    /css-transforms-1_dev/html/reference/transform-abspos-ref.htm 78d197606924062e8dd2a773c977afcecf8940f8
    Testing 71f0313eedfbfcce0b0fcc5ae55f34b1daa3b8d8 == 78d197606924062e8dd2a773c977afcecf8940f8

Wafflespeanut commented:
@bors-servo retry #11561

bors-servo commented:
:zap: Previous build results for [arm32](http://build.servo.org/builders/arm32/builds/2389), [arm64](http://build.servo.org/builders/arm64/builds/2401), [linux-dev](http://build.servo.org/builders/linux-dev/builds/2449), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/2479), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/2462), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/2541), [windows-dev](http://build.servo.org/builders/windows-dev/builds/346) are reusable. Rebuilding only [linux-rel](http://build.servo.org/builders/linux-rel/builds/2579)...

bors-servo commented:
:snowman: The build was interrupted to prioritize another pull request.

bors-servo commented:
:zap: Previous build results for [arm32](http://build.servo.org/builders/arm32/builds/2389), [arm64](http://build.servo.org/builders/arm64/builds/2401), [linux-dev](http://build.servo.org/builders/linux-dev/builds/2449), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/2479), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/2462), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/2541), [windows-dev](http://build.servo.org/builders/windows-dev/builds/346) are reusable. Rebuilding only [linux-rel](http://build.servo.org/builders/linux-rel/builds/2581)...

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/2584)

highfive commented:
▶ Unexpected subtest result in /html/browsers/browsing-the-web/history-traversal/browsing_context_name.html:
      └ PASS [expected FAIL] Retaining window.name on history traversal

ConnorGBrewster commented:
@bors-servo retry
 * #12568

bors-servo commented:
:hourglass: Testing commit 9a3b61a with merge e952040...

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/2586)

highfive commented:
▶ Unexpected subtest result in /html/browsers/browsing-the-web/history-traversal/browsing_context_name.html:
      └ PASS [expected FAIL] Retaining window.name on history traversal

UK992 committed:
Add support colored text in Windows Console

UK992 commented:
Last commit add support for colored text via Windows Console.
Before:
![screenshot 2016-08-13 03 10 42](https://cloud.githubusercontent.com/assets/1530353/17640670/38e79628-6105-11e6-9c44-4dbeed6d7017.png)

After:
![screenshot 2016-08-13 03 11 06](https://cloud.githubusercontent.com/assets/1530353/17640672/3b144af4-6105-11e6-9ad8-e5644ca9b3df.png)

Wafflespeanut commented:
Thanks again! :)

@bors-servo r+

bors-servo commented:
:pushpin: Commit fed0e94 has been approved by `Wafflespeanut`

<!-- @bors-servo r=Wafflespeanut fed0e94ec6d4ac630fa641e32644a7e055e2e6f0 -->

bors-servo commented:
:hourglass: Testing commit fed0e94 with merge 4b5f859...

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/2610)

highfive commented:
▶ FAIL [expected PASS] /css-transforms-1_dev/html/transform-table-007.htm
      └   → /css-transforms-1_dev/html/transform-table-007.htm a5c014b20ef1363bea6f24eda28c7efb7c45698a
    /css-transforms-1_dev/html/reference/transform-blank-ref.htm fa6407b1acbbfea27e27061e7d1bdeca98e4a728
    Testing a5c014b20ef1363bea6f24eda28c7efb7c45698a == fa6407b1acbbfea27e27061e7d1bdeca98e4a728

Wafflespeanut commented:
@bors-servo retry #11574

bors-servo commented:
:hourglass: Testing commit fed0e94 with merge 8419f96...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/2575)

Wafflespeanut commented:
@bors-servo retry #12797

bors-servo commented:
:zap: Previous build results for [arm32](http://build.servo.org/builders/arm32/builds/2419), [arm64](http://build.servo.org/builders/arm64/builds/2432), [linux-dev](http://build.servo.org/builders/linux-dev/builds/2482), [linux-rel](http://build.servo.org/builders/linux-rel/builds/2617), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/2511), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/2493), [windows-dev](http://build.servo.org/builders/windows-dev/builds/377) are reusable. Rebuilding only [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/2575)...

bors-servo commented:
:sunny: Test successful - [arm32](http://build.servo.org/builders/arm32/builds/2419), [arm64](http://build.servo.org/builders/arm64/builds/2432), [linux-dev](http://build.servo.org/builders/linux-dev/builds/2482), [linux-rel](http://build.servo.org/builders/linux-rel/builds/2617), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/2511), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/2493), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/2578), [windows-dev](http://build.servo.org/builders/windows-dev/builds/377)
